[Draft] docs: add VStorage documentation (#1110)
Agoric/documentation: main/reference/vstorage-ref.md

> # VStorage

Reviewer: As the "See also" list at the bottom of the page shows, we have more than one page on VStorage. The heading should distinguish this one somehow. If it's not reference docs for vstorage, what is it?

Do we move reference docs for vstorage elsewhere? Or do we just not have it any more?

In the Agoric platform, VStorage is a key-value store that:

> - provides a **read-only interface where clients of the consensus layer** can query the stored data using specific paths.

Reviewer 2: maybe want to clarify here, "read-only" to external, but contracts can write to it. something like this, so they don't think vstorage is read-only in general.

- is organized in a **hierarchical, path-based structure**. Clients can query these paths to retrieve the data. For example:
```sh
$ agd query vstorage keys 'published.vaultFactory.managers.manager0.vaults'
children:
- vault0
```

> - can be written through a specialized API called **[chainStorage](https://docs.agoric.com/guides/zoe/pub-to-storage.html#publishing-to-chainstorage)** from within the JavaScript VM.

Author: fixed it.

> ![vstorage query diagram](../guides/getting-started/assets/vstorage-brand-q.svg)

Reviewer: This page now has a lot of overlap with `/guides/getting-started/contract-rpc.html`, so there should be links in both directions to clarify how they relate. Perhaps reduce the vstorage material in `contract-rpc`?

Author: The idea is to create a centralized reference for VStorage. I feel centralized reference is okay like this without reducing context at some place else.

Reviewer: Expanding the scope of this page from [reference](https://docs.divio.com/documentation-system/reference/) to also cover [explanation](https://docs.divio.com/documentation-system/explanation/)/[how-to](https://docs.divio.com/documentation-system/how-to-guides/) seems awkward, especially since it overlaps with several other discussions of vstorage.

children:
- vault0
```
For more information on [agd](https://docs.agoric.com/guides/agoric-cli/agd-query-tx), you can refer to the documentation for [query command](https://docs.agoric.com/guides/agoric-cli/agd-query-tx#query-commands) in `agd` CLI.

## VStorage with CLI
VStorage can be accessed via CLI using:

```sh
$ agd [--node {url}] query vstorage {path}
```

<!-- TODO: Add reference to --node documentation-->

For example:
```sh
$ agd query vstorage path published.agoricNames
children:
- brand
- installation
- instance
...
```

With [Agoric CLI](https://docs.agoric.com/guides/agoric-cli/#agoric-cli-reference), you can also use `follow` command to support vstorage query along with some of the marshalling conventions discussed below:
```sh
$ agoric follow -lF :published.agoricNames.brand
[
  [
    "BLD",
    slotToVal("board0566","Alleged: BLD brand"),
  ],
  [
    "IST",
    slotToVal("board0257","Alleged: IST brand"),
  ],
...
]
```

## VStorage Protobuf Service Definition
The protobuf definition of vstorage is [agoric.vstorage.Query](https://github.com/Agoric/agoric-sdk/blob/mainnet1B/golang/cosmos/proto/agoric/vstorage/query.proto#L11) package.
```

> service Query {

Reviewer 2: good idea to put the proto here

> ## VStorage Query API
> VStorage querier service can directly be incorporated in dApps directly using [agoric.vstorage.Query](https://github.com/Agoric/agoric-sdk/blob/mainnet1B/golang/cosmos/proto/agoric/vstorage/query.proto#L11) package.

Reviewer: directly incorporated how?

I don't think we have docs showing how to make / use gRPC clients.

Citing the protobuf definition is useful, but I suggest not going beyond that:

```suggestion
## VStorage Protobuf Service Definition
The protobuf definition is of vstorage is [agoric.vstorage.Query](https://github.com/Agoric/agoric-sdk/blob/mainnet1B/golang/cosmos/proto/agoric/vstorage/query.proto#L11).
```

p.s. "can directly be incorporated" is [passive voice](https://en.wikipedia.org/wiki/Passive_voice). Avoid passive voice unless it's important to emphasize the receiver of the action.

Author: fixed it.

  // Return an arbitrary vstorage datum.
  rpc Data(QueryDataRequest) returns (QueryDataResponse) {
    option (google.api.http).get = "/agoric/vstorage/data/{path}";
  }

  // Return the children of a given vstorage path.
  rpc Children(QueryChildrenRequest)
    returns (QueryChildrenResponse) {
      option (google.api.http).get = "/agoric/vstorage/children/{path}";
  }
}
```

## VStorage chainStorage API
`chainStorage` API offers write-access to VStorage for contracts, by [connecting to a subscriber](https://docs.agoric.com/guides/js-programming/notifiers.html) to a `chainStorage` node.

While adding [Adding Parameter Governance to a Contract](https://docs.agoric.com/guides/governance/#adding-parameter-governance-to-a-contract), `storageNode` and `marshaller` are passed to the contract in its `privateArgs` so it can publish to chainStorage.

```
const marshaller = await E(board).getPublishingMarshaller();
const storageNode = await E(chainStorage).makeChildNode(contractName);
...
const installation = await E(zoe).startInstance(
  ...,
  privateArgs: harden({
      storageNode,
      marshaller,
  })
)
```

> The `chainStorage` node corresponds to the `published` key in the [vstorage hierarchy](https://docs.agoric.com/reference/vstorage-ref.html#vstorage-hierarchy). Using `E(chainStorage).makeChildNode(contractName)` gives the contract access to write to the `published.{contractName}` key and all keys under it. 

Reviewer 2: maybe put a security warning here to ensure contracts aren't passed the top most storage node, because then they can write to any key under it. I made this mistake in the past. 

could be something like 
```
Note: Always ensure to only pass the storage node to your contract that you explicitly intend for it to use. This will avoid giving a contract too much write powers to write to unintended storage nodes.
```

What do you think @ansabgillani ?

To understand more on `Marshaller` and `chainStorage` Node, please refer to documentation on [Publishing to chainStorage](https://docs.agoric.com/guides/zoe/pub-to-storage.html#publishing-to-chainstorage)

## VStorage Viewer
To visualize the current state of VStorage, The [vstorage-viewer](https://github.com/p2p-org/p2p-agoric-vstorage-viewer) contributed by p2p is often _very_ handy:

[![vstorage viewer screenshot](https://user-images.githubusercontent.com/150986/259798595-40cd22f0-fa01-43a9-b92a-4f0f4813a4f6.png)](https://p2p-org.github.io/p2p-agoric-vstorage-viewer/#https://devnet.rpc.agoric.net/|published,published.agoricNames|)

## DApp UI and VStorage
As mentioned above, VStorage offers clients an access to data from contracts, including dApp UI. For detail tutorial on tools and libraries available for this connection is available [here](https://docs.agoric.com/guides/getting-started/ui-tutorial/querying-vstorage.html#querying-vstorage).

> ## Common VStorage Examples

Reviewer: The docs on this page were intended as not just common examples but an exhaustive reference of what's in vstorage, as part of #947.

If this PR is changing that, we would need to renegotiate with the folks involved in #947.

> # VStorage in Agoric

Reviewer 3: remove " in Agoric"

In the Agoric platform, VStorage, or a Virtual Storage, is a key-value store that:

> - provides a **read-only interface for clients of the consensus layer**. Clients can query the stored data using specific paths.

Reviewer 3: change to something like: "...interface where clients can query..."

> - is organized in a **hierarchical, path-based structure**. Each fragment of data is stored at a specific path, and clients can query these paths to retrieve the data.

Reviewer 3: remove this: Each fragment of data is stored at a specific path, and

> - can be written through a specialized API called **[chainStorage](https://docs.agoric.com/guides/zoe/pub-to-storage.html#publishing-to-chainstorage)**. This API allows the VM to manage data storage in a controlled and secure manner.

Reviewer 3: remove: This API allows the VM to manage data storage

![vstorage query diagram](../guides/getting-started/assets/vstorage-brand-q.svg)

## VStorage Hierarchy

VStorage is structured as a tree with paths and nodes that store the actual data. This design facilitates easy querying while ensuring data consistency and security.

For example, here is an example of [Inter Protocol key structure](https://github.com/Agoric/agoric-sdk/tree/agoric-upgrade-13/packages/inter-protocol#reading-data-off-chain) inside Agoric:

```
- published
    - vaultFactory
        - governance
        - metrics
        - managers
            - manager0
                - metrics
                - governance
                - vaults
                    - vault0
```

You can then use this hierarchy to access the data fragments.
```sh
# lists vaults
$ agd query vstorage keys 'published.vaultFactory.managers.manager0.vaults'
children:
- vault0

> ```

Reviewer 3: This is good.

Reviewer: It assumes knowledge of the `agd` CLI. Perhaps add a cross-reference to https://docs.agoric.com/guides/agoric-cli/agd-query-tx.html#query-commands ?

Also, add a section about `agd query vstorage` in https://docs.agoric.com/guides/agoric-cli/agd-query-tx.html ?

Reviewer: Do we expect our readers to be more familiar with `agd` or with JS APIs?

https://docs.agoric.com/guides/getting-started/ui-tutorial/querying-vstorage.html has an example using `chainStorageWatcher.watchLatest` to access `published.agoricNames.instance`. Did you consider using that?

Author: I understand. I have added both, JS APIs and `agd` here.

Author: Added here #1122

Reviewer: The JS API that has been added is how to _write_ to vstorage / chainstorage from a contract. I was suggesting adding the API for _querying_ vstorage from a client dapp, as suggested in #947. I suppose both are relevant. I'm not sure how big the scope should grow here, though.

## VStorage with CLI
VStorage can be accessed via CLI using:

```sh
$ agd /[--node {url}] query vstorage {path}
```

For example:
```sh
$ agd query vstorage path published.agoricNames
children:
- brand
- installation
- instance

> ...

Reviewer: Ah. yes. this is the sort of material that I'd like to see in https://docs.agoric.com/guides/agoric-cli/agd-query-tx.html

The `--node` flag is discussed more thoroughly there. I don't see any reference to it here, so you might as well drop it here.

What's the `/` in `/[--node {url}]` by the way?

Author: `/` was a typo. Fixing it in a commit. 

```
The query goes to a local node at tcp://localhost:26657 by default. To use another node:
$ agd status --node https://devnet.rpc.agoric.net:443
```

Can you confirm this is same for `query` as well? I am not sure about it. There is no mention about `--node` parameter there. Maybe I can add information there and then refer it to that.

Reviewer: yes, `--node` applies to all `agd query` and `agd tx` sub-commands, IME.

The help text is oddly inconsistent, though: `agd query --help | grep node` is empty, as is `agd query vstorage --help | grep node` but:

```console
$ agd query tx --help | grep node
      --node string        <host>:<port> to Tendermint RPC interface for this chain (default "tcp://localhost:26657")
```

@gibson042 Do you know of some reason why the help text is inconsistent? If not, I suppose we owe an issue?

Reviewer 4: Yes, an issue would be good. I'm pretty sure this is because the [Cobra](https://cobra.dev/) library responsible for CLI option definition shows help for applicable options defined higher in the command hierarchy (so-called ["persistent flags"](https://cobra.dev/#persistent-flags)) only at an invocable command (which `agd query tx` is but e.g. `agd query vstorage` is not because it must be followed by "children" or "data" or "path" and `agd query params` is not because it must be followed by "subspace").

But there should be a way to do it, because "--chain-id" seems to work that way (e.g., it shows in both e.g. `agd tx --help` and [as a Global Flag] in `agd tx gov --help`).

```

<!-- mention the constant marshalling in follow --> 
With Agoric CLI, you can also use `follow` command to support vstorage query plus some of the marshalling conventions discussed below:
```sh
$ agoric follow -lF :published.agoricNames.brand
[
  [
    "BLD",
    slotToVal("board0566","Alleged: BLD brand"),
  ],
  [
    "IST",
    slotToVal("board0257","Alleged: IST brand"),
  ],
...
]
```

> 

Reviewer 3: This does not whole lot of sense - let us get rid of this.

Author: I have removed it for now. This is important. Already referred here and we can get extra info on this from 
https://fa4fcd61.documentation-7tp.pages.dev/guides/agoric-cli/#agoric-help

## VStorage Query API

## VStorage chainStorage API

> ## VStorage Viewer

Reviewer: These headings are taken from an semi-automated test. see the commit message for https://github.com/Agoric/documentation/commit/5607e3bf139bbe90aefd93b227e48a53a5a096e7 for details

I didn't manage to put the test into ci in such a way that changing the heading here without changing the test would be flagged as an inconsistency. I wonder what it would take to do that.

Author: 1. I will revert these changes. I understand. 
2. Maybe run a script that opens the MD file and imports the JSON object with headings from the test and compares the two. If the headings are dependent on the result of a test, we can add it as a subsequent step in the CI.

    }
```

### Namehubs - agoricNames hubs

agoricNames contains several other NameHubs.
See also [agoricNames](https://docs.agoric.com/guides/integration/name-services.html#agoricnames-agoricnamesadmin-well-known-names).

```js
[
  'brand',
  'installation',
  'instance',
  'issuer',
  'oracleBrand',
  'vbankAsset'
];
```

### Well known contracts

`published.agoricNames.installation` contains _Installations_ representing code of important contracts. The data at this key are the entries of the NameHub. Here we show the object comprised
of those entries.
See also [agoricNames in vstorage](https://docs.agoric.com/guides/integration/name-services.html#agoricnames-in-vstorage)
regarding un-marshalling the data using board IDs.

```js
{
  VaultFactory: Object @Alleged: BundleIDInstallation#board05815 {},

>   auctioneer: Object @Alleged: BundleIDInstallation#board04016 {},

Reviewer: likewise, these data were produced by a semi-automated test.

Is it important to change the indentation?

Author: I did it via linter. Reverting these changes.

      published: 'for the chainStorage API; see below',
    }
```

> ### vstorage: published.\* keys

Reviewer: Demoting these from h2 to h3 means they don't show up in the page TOC. Is that on purpose?

`published.agoricNames.instance` contains _instances_ of important contracts.
The data at this key are the entries of the NameHub. Here we show the object comprised of those entries.

`published.agoricNames.instance` contains _instances_ of important contracts.
The data at this key are the entries of the NameHub. Here we show the object comprised of those entries.

> 

Reviewer: I don't think you meant to repeat this.

          },
        ],
      },
      brand: Object @Alleged: Zoe Invitation brand#board0074 {},
    },
  ],
}
```

The `published.wallet.${address}` key has wallet's last update. For example:

```js
{
  currentAmount: {
    brand: Object @Alleged: Zoe Invitation brand#board0074 {},
    value: [
      {
        description: 'Voter0',
        handle: Object @Alleged: InvitationHandle#null {},
        installation: Object @Alleged: BundleIDInstallation#board00613 {},
        instance: Object @Alleged: InstanceHandle#null {},
      },
      {
        description: 'charter member invitation',
        handle: Object @Alleged: InvitationHandle#null {},
        installation: Object @Alleged: BundleIDInstallation#board01679 {},
        instance: Object @Alleged: InstanceHandle#null {},
      },
    ],
  },
  updated: 'balance',
}
```

See also:

> - [Querying VStorage](/guides/getting-started/contract-rpc#querying-vstorage)

Reviewer: https://docs.agoric.com/guides/getting-started/ui-tutorial/querying-vstorage.html should probably go before `/getting-started/contract-rpc` in this see also list
